### PR TITLE
feat: Implement release-based contract deployment

### PR DESCRIPTION
**Pull request by blmalone**
Added functionality to either reuse previously deployed implementations of deploy new implementations based on the release string that's passed through to `DeployImplementations.s.sol`. 

`op-deployer` passes the `standard-versions.toml` string though to `DeployImplementations.s.sol` as input. A foundry cheatcode for  reading toml is reimplemented in go to make this work. 

Follow up work is needed to make this a cleaner solution, which is outlined in this issue: https://github.com/ethereum-optimism/optimism/issues/12067

### CODE REVIEW

**Comment by blmalone** [comment]
* **#12035** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/12035?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @blmalone and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/contracts-bedrock/scripts/DeployImplementations.s.sol
Comment (blmalone) [inline packages/contracts-bedrock/scripts/DeployImplementations.s.sol:499]: Looks like we're following the convention everywhere so adding it to these functions too.

File: packages/contracts-bedrock/scripts/DeployImplementations.s.sol
Comment (blmalone) [inline packages/contracts-bedrock/scripts/DeployImplementations.s.sol]: Notice how because mips is a singleton it has its key as [address](https://github.com/ethereum-optimism/superchain-registry/blob/be65d22f8128cf0c4e5b4e1f677daf86843426bf/validation/standard/standard-versions.toml#L19) whereas the dispute game factory is proxied and has its key set as [implementation_address](https://github.com/ethereum-optimism/superchain-registry/blob/be65d22f8128cf0c4e5b4e1f677daf86843426bf/validation/standard/standard-versions.toml#L16).

File: packages/contracts-bedrock/test/DeployImplementations.t.sol
Comment (blmalone) [inline packages/contracts-bedrock/test/DeployImplementations.t.sol:342]: this test proves that you won't accidentally deploy implementations if the `op-contracts/*` tag input by a user does not exist.

File: packages/contracts-bedrock/test/DeployImplementations.t.sol
Comment (blmalone) [inline packages/contracts-bedrock/test/DeployImplementations.t.sol:397]: this test proves that you won't accidentally deploy implementations if the `op-contracts` tag exists but the contract doesn't.

File: op-chain-ops/deployer/integration_test/apply_test.go
Comment (blmalone) [inline op-chain-ops/deployer/integration_test/apply_test.go:118]: This test _does not_ use `standard-versions.toml` file.

File: op-chain-ops/Makefile
Comment (mds1) [inline op-chain-ops/Makefile:49]: When is this intended to be ran? Can we add some info to the PR description detailing how to use this new feature?

File: op-chain-ops/deployer/opsm/standard-versions.toml
Comment (mds1) [inline op-chain-ops/deployer/opsm/standard-versions.toml:1]: Why do we commit a file that already exists in the superchain registry? I worry about this getting out of date as it's another step to maintain for future OPCM versions. Should we instead gitignore this file to ensure the `sync-standard-version` command is always run?

File: op-chain-ops/script/cheatcodes_utilities.go
Comment (mds1) [inline op-chain-ops/script/cheatcodes_utilities.go:197]: Why the `_65e7c844` suffix?

File: packages/contracts-bedrock/scripts/DeployImplementations.s.sol
Comment (mds1) [inline packages/contracts-bedrock/scripts/DeployImplementations.s.sol:65]: Based on the above, this is supposed to be the content itself, and not a path to it right? What is the rationale for that?

File: op-chain-ops/deployer/opsm/standard-versions.toml
Comment (mslipper) [reply op-chain-ops/deployer/opsm/standard-versions.toml:1]: OP Deployer is meant to be a standalone library with minimal outside dependencies apart from the intent, state file, and contract artifacts. The way the OPCM is currently implemented, we need to provide the contents of the standard versions TOML to the OPCM for it to know where the implementation addresses are. I don't want users to have to include this file all the time, nor do I want them to have to download it on first run of the op-deployer tool. So, to get around this limitation, the TOML is fetched from the registry and stored in the monorepo. It is then embedded in the OP Deployer binary via `goembed`, and provided as a string to the OPCM.

In the future, I'd rather generate a Solidity contract that can be included with the bytecode from the superchain registry or generate structs in the monorepo that encode the TOML files themselves. Since that would require a rewrite of Blaine's work, we opted for what you see here.

File: op-chain-ops/script/cheatcodes_utilities.go
Comment (mslipper) [reply op-chain-ops/script/cheatcodes_utilities.go:197]: Go doesn't support overloads, so the suffix is the 4byte selector for this specific overload.

File: packages/contracts-bedrock/scripts/DeployImplementations.s.sol
Comment (mslipper) [reply packages/contracts-bedrock/scripts/DeployImplementations.s.sol:65]: See my comment [above](https://github.com/ethereum-optimism/optimism/pull/12035#discussion_r1772258525).

File: op-chain-ops/Makefile
Comment (mslipper) [reply op-chain-ops/Makefile:49]: We'll need to run it whenever the standard versions change.

File: op-chain-ops/deployer/opsm/standard-versions.toml
Comment (mslipper) [reply op-chain-ops/deployer/opsm/standard-versions.toml:1]: The TOML file must be present in the repo at op-deployer's compile time for the goembed to work, so we're committing it here.

File: packages/contracts-bedrock/test/DeployImplementations.t.sol
Comment (mds1) [inline packages/contracts-bedrock/test/DeployImplementations.t.sol:371]: Not quite following when we'd uncomment this

File: packages/contracts-bedrock/test/DeployImplementations.t.sol
Comment (mds1) [reply packages/contracts-bedrock/test/DeployImplementations.t.sol:397]: Great test cases 👌

File: packages/contracts-bedrock/scripts/DeployImplementations.s.sol
Comment (mds1) [reply packages/contracts-bedrock/scripts/DeployImplementations.s.sol]: This makes sense to me! Resolving

File: packages/contracts-bedrock/test/DeployImplementations.t.sol
Comment (blmalone) [reply packages/contracts-bedrock/test/DeployImplementations.t.sol:371]: Should Op contracts manager implementation be included in the superchain registry at some stage? https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/standard/standard-versions.toml

If so, this code would be uncommented. Right now, in `DeployImplementation.s.sol` I'm deploying Op contracts manager impl fresh everytime.

File: op-chain-ops/deployer/opsm/standard-versions.toml
Comment (mds1) [reply op-chain-ops/deployer/opsm/standard-versions.toml:1]: What exactly would this solidity contract or structs look like? I don't fully follow the idea and longer-term plan described in https://github.com/ethereum-optimism/optimism/issues/12067

Either way, thanks for the explanation! This is ok with me for now

File: packages/contracts-bedrock/test/DeployImplementations.t.sol
Comment (mds1) [reply packages/contracts-bedrock/test/DeployImplementations.t.sol:371]: Hmm good question, will need to think on that. I can see arguments for both. For now I like this TODO, and I added a sticky to the figjam to track this

**Comment by mds1** [comment]
This LGTM, just holding off on approving so we ensure https://github.com/ethereum-optimism/optimism/pull/12064 is merged before this one

~~Failing CI seems to be an out of order nonce issue when broadcasting?~~ Looks like it was just a flaky test

File: op-chain-ops/deployer/opsm/standard-versions.toml
Comment (mds1) [reply op-chain-ops/deployer/opsm/standard-versions.toml:1]: Moved my question into that issue to resolve this thread

File: packages/contracts-bedrock/scripts/DeployImplementations.s.sol
Comment (mds1) [inline packages/contracts-bedrock/scripts/DeployImplementations.s.sol]: One thing I don't love is that we `startBroadcast` then jump into a function. So when modifying that `deployOPContractsManagerImpl` function the dev hsa to remember they are in a broadcast context, which isn't obvious. I would suggest changing this `vm.startBroadcast` to just `vm.broadcast`, and again using `vm.broadcast` within `deployOPContractsManagerImpl`. Then for the below calls, you can either use multiple `vm.broadcast` calls or bundle them within a start/stopBroadcast

File: packages/contracts-bedrock/scripts/DeployImplementations.s.sol
Comment (mds1) [inline packages/contracts-bedrock/scripts/DeployImplementations.s.sol]: same comment around `broadcast` vs. `startBroadcast`

**Review by mds1 — APPROVED** [review]

File: packages/contracts-bedrock/scripts/DeployImplementations.s.sol
Comment (blmalone) [reply packages/contracts-bedrock/scripts/DeployImplementations.s.sol]: Good call, added in:https://github.com/ethereum-optimism/optimism/pull/12035/commits/c131a31a0cd4683f239d8a59df7fc6bd70f271ed

File: packages/contracts-bedrock/scripts/DeployImplementations.s.sol
Comment (blmalone) [reply packages/contracts-bedrock/scripts/DeployImplementations.s.sol]: Added in: https://github.com/ethereum-optimism/optimism/pull/12035/commits/c131a31a0cd4683f239d8a59df7fc6bd70f271ed

**Review by mslipper — APPROVED** [review]

**Pull request merged**
